File: zaptorch-base/src/main/java/com/pyamsoft/zaptorch/base/ZapTorchModule.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.zaptorch.base;

import android.app.IntentService;
import android.content.Context;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import rx.Scheduler;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class ZapTorchModule {

  @NonNull private final Context appContext;
  @NonNull private final ZapTorchPreferences preferences;
  @NonNull private final Class<? extends IntentService> torchOffServiceClass;

  public ZapTorchModule(@NonNull Context context,
      @NonNull Class<? extends IntentService> torchOffServiceClass) {
    appContext = context.getApplicationContext();
    preferences = new ZapTorchPreferencesImpl(context);
    this.torchOffServiceClass = torchOffServiceClass;
  }

  @CheckResult @NonNull public Context provideContext() {
    return appContext;
  }

  @CheckResult @NonNull public ZapTorchPreferences providePreferences() {
    return preferences;
  }

  @CheckResult @NonNull public Class<? extends IntentService> provideTorchOffServiceClass() {
    return torchOffServiceClass;
  }

  @CheckResult @NonNull public Scheduler provideObsScheduler() {
    return AndroidSchedulers.mainThread();
  }

  @CheckResult @NonNull public Scheduler provideSubScheduler() {
    return Schedulers.io();
  }
}

File: zaptorch-base/src/main/java/com/pyamsoft/zaptorch/base/ZapTorchModule.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.zaptorch.base

import android.app.IntentService
import android.content.Context
import android.support.annotation.CheckResult
import com.pyamsoft.pydroid.base.PYDroidModule
import com.pyamsoft.pydroid.loader.ImageLoader
import com.pyamsoft.pydroid.loader.LoaderModule
import com.pyamsoft.zaptorch.base.preference.CameraPreferences
import com.pyamsoft.zaptorch.base.preference.ClearPreferences
import com.pyamsoft.zaptorch.base.preference.UIPreferences
import io.reactivex.Scheduler

class ZapTorchModule(private val pyDroidModule: PYDroidModule,
        private val loaderModule: LoaderModule,
        private val torchOffServiceClass: Class<out IntentService>) {

    private val preferences: ZapTorchPreferencesImpl = ZapTorchPreferencesImpl(
            pyDroidModule.provideContext())

    @CheckResult
    fun provideContext(): Context = pyDroidModule.provideContext()

    @CheckResult
    fun provideCameraPreferences(): CameraPreferences = preferences

    @CheckResult
    fun provideClearPreferences(): ClearPreferences = preferences

    @CheckResult
    fun provideUiPreferences(): UIPreferences = preferences

    @CheckResult
    fun provideTorchOffServiceClass(): Class<out IntentService> = torchOffServiceClass

    @CheckResult
    fun provideMainThreadScheduler(): Scheduler = pyDroidModule.provideMainThreadScheduler()

    @CheckResult
    fun provideIoScheduler(): Scheduler = pyDroidModule.provideIoScheduler()

    @CheckResult
    fun provideComputationScheduler(): Scheduler = pyDroidModule.provideComputationScheduler()

    @CheckResult
    fun provideImageLoader(): ImageLoader = loaderModule.provideImageLoader()
}
